src/XAD/Hessian.hpp

```diff
@@ -0,0 +1,74 @@
+/*******************************************************************************
+
+   Implementation of Hessian computing methods.
+
+   This file is part of XAD, a comprehensive C++ library for
+   automatic differentiation.
+
+   Copyright (C) 2010-2024 Xcelerit Computing Ltd.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU Affero General Public License as published
+   by the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+******************************************************************************/
+
+#include <functional>
+#include <vector>
+
+namespace xad
+{
+
+template <class T>
+class Hessian
+{
+  public:
+    Hessian(std::function<T(std::vector<T> &)> func, std::vector<T> &v) : foo(func), v(v) {}
+
+    std::vector<std::vector<T>> compute()
```

Comment:
wondering if we shouldn't support more generic matrix implementations (some users are using Eigen for example). We could for example have a function:

```c++

template <class RowIterator>
void compute(RowIterator first, RowIterator last) {
}
```

With the constraint that each row iterator has a `.begin()` member that provides an element iterator along that row? Given that the Hessian is symmetric, it doesn't even matter if it's a row or a column iterator. 

Then this function could be implemented on top of it for convenience:

```
std::vector<std::vector<T>> compute() {
  domain = static_cast<unsigned int>(v.size());
  std::vector<std::vector<T>> matrix(domain, std::vector<T>(domain, 0.0));
  compute(matrix.begin(), matrix.end());
  return matrix;
}
```

and it would be straightforward to use with other storage formats. Note that some range checks should probably be performed in debug mode (`std::distance(first, last) == v.size()`) and the like.

Reply:
Hessian matrices are only symmetrical if the second derivatives are continuous iirc. Alternatively, I could use a template and as long as the datatype passed defines operator[] then it should work I reckon.

Reply:
You're right right re symmetry in general. But if we clearly state that it is a *row* iterator that is no issue. Whether that is equivalent to a column iterator depends on symmetry - and users need to be aware when they use the iterator interface and pass a column iterator instead.

We prefer using an output iterator over `operator[]`, as that would be more generic and there's no need to restrict the type further in this context. (`operator[]` would fail for `list<vector<T>>` for example, while the iterator approach would allow that)

Reply:
Added overloads for `RowIterator`.

Am I free to add exceptions for:
- Iterator distance < domain_*codomain_
- Iterator doesn't support begin()

Reply:
Regarding the first one, it makes sense to add that. It's a range error - I think the `OutOfRange` exception that's already there makes sense. 

For the second one, I think that can be statically (compile-time assert). Using the tools in `<type_traits>`, it should be possible to check if the iterator's value-type has the method `begin` at compile-time and raise an assert. In C++20, we could use concepts, but I wouldn't go there yet.